### PR TITLE
Completed adding cache to component, improved asynchronous loading

### PR DESCRIPTION
**Pull request by MarkPThomas**
More to come, but this needs to be merged first!

### CODE REVIEW

File: client/src/components/overview/ProductOverview.jsx [hunk 1]
```diff
@@ -88,25 +121,35 @@ class ProductOverview extends React.Component {
...
   }
 
-  getDefaultStyle() {
-    for (const style of this.styles) {
+  updateRatingsProperties(ratings) {
+    console.log(`Product Ratings: ${JSON.stringify(ratings)}`);
+    const numberOfReviews = this.getNumberOfReviews(ratings);
+    this.setState({
+      numberOfReviews: numberOfReviews,
+      averageRating: this.getAverageRating(ratings, numberOfReviews)
```
Comment (stevenharderjr) [inline client/src/components/overview/ProductOverview.jsx:135]: Why call `getAverageRating` in `setState` but assigning a temp variable for `getNumberOfRatings`?

File: client/src/components/overview/ProductOverview.jsx [hunk 2]
```diff
@@ -88,25 +121,35 @@ class ProductOverview extends React.Component {
...
         return style;
       }
     }
     console.log('No default found! Using first style...');
-    return this.styles.length > 0 ? this.styles[0] : null;
+    return styles.length > 0 ? styles[0] : null;
   }
 
-  getStyleById(id) {
+  getStyleById(id, styles) {
```
Comment (stevenharderjr) [inline client/src/components/overview/ProductOverview.jsx:150]: `checkCache` and `updateCache` use a Map object to performs this sort of operation in constant time.

File: client/src/components/overview/ProductOverview.jsx [hunk 3]
```diff
@@ -118,16 +161,17 @@ class ProductOverview extends React.Component {
   setStyleById(id) {
     id = parseInt(id);
     if (this.state.selectedStyle.style_id !== id) {
-      const style = this.getStyleById(id);
+      const style = this.getStyleById(id, this.state.styles);
```
Comment (stevenharderjr) [inline client/src/components/overview/ProductOverview.jsx:164]: I know you're currently refactoring, but unless you have some surprises coming, `styles` does not seem like something that React state is set up to deal with. Keep in mind, setState is a rather unpredictable async operation. Unless you need to trigger renders, using the cache or regular class variables (`this.styles`) and regular assignment might be a better/less error prone way to keep track of things.

File: client/src/components/overview/ProductOverview.jsx [hunk 4]
```diff
@@ -118,16 +161,17 @@ class ProductOverview extends React.Component {
...
+      // TODO: Above can be refactored into model that takes in styles, selectedStyleId, and id
       if (style) {
         this.updateState(style);
       }
     }
   }
 
-  getStyleSelectorItems() {
+  getStyleSelectorItems(styles) {
```
Comment (stevenharderjr) [inline client/src/components/overview/ProductOverview.jsx:172]: `styleSelectorItems` seem like they should be components.

File: client/src/components/overview/ProductOverview.jsx [hunk 5]
```diff
@@ -8,16 +8,19 @@ import ProductDescription from './ProductDescription';
...
+  //   updateCache
+  // }
     super(props);
 
-    this.numberOfReviews = 0;
-    this.averageRating = 0;
-    this.styles = [];
-
     this.state = {
+      styles: [],
```
Comment (stevenharderjr) [inline client/src/components/overview/ProductOverview.jsx:19]: I would think more about which of these actually need to be tracked, and how well-suited they are to being tracked as component state.

File: client/src/components/overview/ProductOverview.jsx [hunk 6]
```diff
@@ -42,35 +45,65 @@ class ProductOverview extends React.Component {
...
-        this.styles = res[1].data.results;
-        this.updateDefaultStyle();
-      })
-      .catch(error => {
-        console.error(error.stack);
-      });
+
+    this.fetchRatings(id, this.updateRatingsProperties);
+    this.fetchProductStyles(id, (styles) => {
+      this.setState({ styles });
```
Comment (stevenharderjr) [inline client/src/components/overview/ProductOverview.jsx:62]: This is unlikely to work the way you want component state to work. If you keep updating the same `styles` object, React will fail to notice and nothing will update. If you are careful to create a new `styles` object every time you set state, React will re-render everything that uses this data, even if some/all of the actual content is the same.

**Review by stevenharderjr — APPROVED** [review]
If this goes unchanged, I think you'll be in for some extraordinarily hard debugging.

File: client/src/components/overview/ProductOverview.jsx (hunk 1 above)
Comment (MarkPThomas) [reply client/src/components/overview/ProductOverview.jsx:135]: It seemed like I was courting problems by using the value of one state to determine another while they are being set together. Cleaner to just calculate it outside of setting state.

File: client/src/components/overview/ProductOverview.jsx (hunk 2 above)
Comment (MarkPThomas) [reply client/src/components/overview/ProductOverview.jsx:150]: Yep, I saw that. A later refactor sweep to do after some other optimizations. For now the scope of this PR was just to stop making REST calls isolated from the rest of the app. I see more refactoring to do in later PRs.

File: client/src/components/overview/ProductOverview.jsx (hunk 3 above)
Comment (MarkPThomas) [reply client/src/components/overview/ProductOverview.jsx:164]: I need to look more broadly later to be certain, but this was necessary once using the cache, as I need to trigger a render at some unknown later time. Before I made this change, the component didn't update later in the app render cycle once data was available.

File: client/src/components/overview/ProductOverview.jsx (hunk 5 above)
Comment (MarkPThomas) [reply client/src/components/overview/ProductOverview.jsx:19]: Agreed. Just getting it to work for now. Before, styles was a local props, but moving the data to cache requires some rethinking of props vs state & where it needs to be.

**Pull request merged**